src/fillalgebra.jl

```diff
@@ -100,23 +118,61 @@ end
 *(a::Transpose{T, <:StridedMatrix{T}}, b::Fill{T, 1}) where T = reshape(sum(parent(a); dims=1) .* b.value, size(parent(a), 2))
 *(a::StridedMatrix{T}, b::Fill{T, 1}) where T         = reshape(sum(a; dims=2) .* b.value, size(a, 1))
 
-function *(a::Adjoint{T, <:StridedMatrix{T}}, b::Fill{T, 2}) where T
-    fB = similar(parent(a), size(b, 1), size(b, 2))
-    fill!(fB, b.value)
-    return a*fB
+function *(x::AbstractMatrix, f::FillMatrix)
+    axes(x, 2) ≠ axes(f, 1) &&
+        throw(DimensionMismatch("Incompatible matrix multiplication dimensions"))
+    m = size(f, 2)
+    repeat(sum(x, dims=2) * f.value, 1, m) 
+end
+
+function *(f::FillMatrix, x::AbstractMatrix)
+    axes(f, 2) ≠ axes(x, 1) &&
+        throw(DimensionMismatch("Incompatible matrix multiplication dimensions"))
+    m = size(f, 1)
+    repeat(sum(x, dims=1) * f.value, m, 1) 
 end
 
-function *(a::Transpose{T, <:StridedMatrix{T}}, b::Fill{T, 2}) where T
-    fB = similar(parent(a), size(b, 1), size(b, 2))
-    fill!(fB, b.value)
-    return a*fB
+function *(x::AbstractMatrix, f::OnesMatrix)
+    axes(x, 2) ≠ axes(f, 1) &&
+        throw(DimensionMismatch("Incompatible matrix multiplication dimensions"))
+    m = size(f, 2)
+    repeat(sum(x, dims=2) * one(eltype(f)), 1, m) 
 end
 
-function *(a::StridedMatrix{T}, b::Fill{T, 2}) where T
-    fB = similar(a, size(b, 1), size(b, 2))
-    fill!(fB, b.value)
-    return a*fB
+function *(f::OnesMatrix, x::AbstractMatrix)
+    axes(f, 2) ≠ axes(x, 1) &&
+        throw(DimensionMismatch("Incompatible matrix multiplication dimensions"))
+    m = size(f, 1)
+    repeat(sum(x, dims=1) * one(eltype(f)), m, 1) 
 end
+
+*(x::FillMatrix, y::FillMatrix) = mult_fill(x, y)
+*(x::FillMatrix, y::OnesMatrix) = mult_fill(x, y)
+*(x::OnesMatrix, y::FillMatrix) = mult_fill(x, y)
+*(x::OnesMatrix, y::OnesMatrix) = mult_fill(x, y)
+*(x::ZerosMatrix, y::OnesMatrix) = mult_zeros(x, y)
+*(x::ZerosMatrix, y::FillMatrix) = mult_zeros(x, y)
+*(x::FillMatrix, y::ZerosMatrix) = mult_zeros(x, y)
+*(x::OnesMatrix, y::ZerosMatrix) = mult_zeros(x, y)
```

Comment:
I think you are massively underestimating the number of ambiguities we need to take into account. For example we need to take into account `AdjointAbsVec`, `TransposeAbsVec`, `AbstractTriangular`, `Adjoint`, `Transpose`, .... A somewhat more exhaustive list is here: 

https://github.com/JuliaMatrices/ArrayLayouts.jl/blob/9fbed3a8e49b26d4ccff5608635b8c986f702b7a/src/mul.jl#L158


It's also a bad idea to overload `Ones` and `Fills` seperatly since that just means more ambiguities downstream. The following is better:
```julia
*(a::AbstractMatrix, b::AbstractFill{<:Any,2}) = fill_mul(a, b)
fill_mul(a, b::Ones) = ...
fill_mul(a, b::Zeros) = ...
```

Reply:
hopefully I addressed both concerns

Reply:
Is there a chance that overloading something other than `*` might be simpler, and avoid all ambiguities? They are pretty thorny, as every other package that tries to play the same game introduces new possibilities...

Without special handling, it will end up after `similar` at some `generic_mul!(C, A::Fill, B)`. Could one overload that instead? I'm not precisely sure where in the chain.

Reply:
This issue is precisely what ArrayLayouts.jl is designed to avoid, and would also give us the ` mul! ` variants. One option would be to  make FillArrays.jl depend on ArrayLayouts.jl (it’s currently the other way around).

Reply:
Wait how does that work? Is this only if you call something special like `ArrayLayouts.mul` or does it stick its fingers into ordinary `A * B`?

Reply:
We would do ` AbstractFill <: LayoutArray ` which then catches all the necessary  ` * ` ovverrides

Reply:
OK, got it. Overloading for one supertype stops all its children from stepping on each others' toes.

Reply:
although that's probably the right direction, It involves a lot of churn and the coordination of two packages, so I would merge this as it is for the time being

Reply:
Yes I agree.

Reply:
> the coordination of two packages

FYI I wrote both packages so this isn't quite so bad. The only reason I haven't done this is FillArrays.jl is clean and simple, so needs a good reason to make it more complicated. I'm not sure we are there yet.